Standardize logging locations across desktop platforms

jcm93 opened the pull request:
### Description
Unifies log directory finding and creation in `AppDataManager`, and ensures logs are created in the intended, correct places on each platform.

The preferred directory hierarchy is the following.

Windows: Prefer a 'Logs' folder in whatever directory Ryujinx was launched from; fall back to `%APPDATA%\Ryujinx\Logs`.

Mac: Prefer `~/Library/Logs/Ryujinx`. Fall back to `~/Library/Application Support/Ryujinx/Logs`.

Linux: Use `~/.config/Ryujinx/Logs`.

### Context

https://github.com/Ryujinx/Ryujinx/commit/70fcba39de34fc211d09da12783898161724b0dc had an unfortunate quality escape where logs on macOS were being written to the executable directory, which is within the .app bundle. This was invalidating the code signature of the app bundle, which can lead to various unintended side effects. This PR addresses that issue while standardizing log locations across Windows, Mac and Linux.

Fix #6252

gdkchan commented:
Why not `C:\Users\user\AppData\Roaming\Ryujinx\Logs` on Windows (as fallback) which is were the emulator config and everything else is stored?

jcm93 commented:
Whoops, I misinterpreted something offthread and thought we wanted %LOCALAPPDATA% instead of %APPDATA%. I'll change it to %APPDATA%.

Steveice10 commented:
This doesn’t appear to account for portable mode, which could make logs more annoying to get to.

jcm93 commented:
Updated to respect portable mode.

jcm93 commented:
There is some discussion still offthread about the appropriate location to log on Windows. Just wanted to put forward my two cents that I hope that discussion is non-blocking for this PR. Most of all, it is important to stop macOS from invalidating its own signature. An invalid code signature can stop Ryujinx from being able to write to its own Application Support folder, stop Ryujinx from being able to load its own libraries, stop it from launching entirely, etc. This also *must* be merged before https://github.com/Ryujinx/Ryujinx/pull/6241, assuming both are merged, since it's very probable that the copy and symlink removal operation there will fail if Ryujinx's code signature is invalid.

jcm93 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,8 +113,116 @@ public static void Initialize(string baseDirPath)
+                    if (logDir == "")
+                    {
+                        //Should evaluate to ~/Library/Application Support/Ryujinx/Logs
+                        logDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
```
Not really, probably better to use BaseDirPath; changed.

jcm93 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -30,6 +30,8 @@ public enum LaunchMode
         public static string KeysDirPath { get; private set; }
         public static string KeysDirPathUser { get; }
 
+        public static string LogsDirPath { get; private set; }
```
Removed. As a result, changed `MainWindowViewModel`'s `OpenLogsFolder()` to use `AppDataManager` as well, rather than LoggerModule.

jcm93 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,8 +113,116 @@ public static void Initialize(string baseDirPath)
+                    {
+                        Logger.Warning?.Print(LogClass.Application,
+                            $"Logging directory could not be created '{logDir}'");
+                        logDir = "";
```
Changed all final exception paths above to return null instead of set logDir to "".

jcm93 commented:
I noticed while addressing feedback that if we pass an empty string or null to `FileLogTarget`'s `PrepareLogFile(string path)`, we have an uncaught exception. I assume we don't want to crash entirely if we can't open the log file, so I might address that in a bit.

jcm93 commented:
Adjusted `FileLogTarget` to catch exception when creating a `DirectoryInfo` for the path. The catch is general; I'm not sure if we can realistically hit anything besides ArgumentException at this point, lemme know if you want to just catch that specifically.

Tested that Ryujinx will now start up properly if it gets a null/empty string when creating the log.

TSRBerry commented:
> I noticed while addressing feedback that if we pass an empty string or null to `FileLogTarget`'s `PrepareLogFile(string path)`, we have an uncaught exception here:
> 
> https://github.com/Ryujinx/Ryujinx/blob/25123232bd7186d8009332f67124a5991f2a978d/src/Ryujinx.Common/Logging/Targets/FileLogTarget.cs#L26
> 
> I assume we don't want to crash entirely if we can't open the log file, so I might address that in a bit.

It was fine before, since we used null checks before calling that function. But yes, a null check / catch statement is better anyway.

TSRBerry commented on src/Ryujinx.Common/Logging/Targets/FileLogTarget.cs:
```diff
@@ -23,7 +23,17 @@ public FileLogTarget(string name, FileStream fileStream)
+            {
+                logDir = new(path);
+            }
+            catch (Exception exception)
+            {
+                Logger.Warning?.Print(LogClass.Application, $"Logging directory path was invalid: {exception}");
```
Catching `ArgumentException` specifically is better imo. I'd also log the invalid path, so we can actually see it in a log.

TSRBerry commented on src/Ryujinx/Ui/MainWindow.cs:
```diff
@@ -1378,9 +1378,9 @@ private void OpenLogsFolder_Pressed(object sender, EventArgs args)
         {
             string logPath = System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Logs");
 
-            if (LoggerModule.LogDirectoryPath != null)
+            if (AppDataManager.LogsDirPath != null)
             {
-                logPath = LoggerModule.LogDirectoryPath;
+                logPath = AppDataManager.LogsDirPath;
             }
```
Same as for Avalonia. We only need to keep `AppDataManager.LogsDirPath` now, since never want to open a different directory here (if it isn't null because then we don't want to open anything).

jcm93 commented:
Addressed, thanks.

TSRBerry reviewed (approved):
LGTM, thank you very much!
I'll just add a commit to address some styling issues (so we avoid stalling this for too long).

gdkchan commented on src/Ryujinx.Common/Logging/Targets/FileLogTarget.cs:
```diff
@@ -23,7 +23,18 @@ public FileLogTarget(string name, FileStream fileStream)
+            }
+            catch (ArgumentException exception)
+            {
+                Logger.Warning?.Print(LogClass.Application, $"Logging directory path ({path}) was invalid: {exception}");
```
```suggestion
                Logger.Warning?.Print(LogClass.Application, $"Logging directory path ('{path}') was invalid: {exception}");
```

gdkchan commented on src/Ryujinx.Ui.Common/Configuration/LoggerModule.cs:
```diff
@@ -9,7 +9,6 @@ namespace Ryujinx.Ui.Common.Configuration
     public static class LoggerModule
     {
-        public static string LogDirectoryPath { get; private set; }
 
```
```suggestion
```

gdkchan commented on src/Ryujinx.Ava/UI/ViewModels/MainWindowViewModel.cs:
```diff
@@ -1350,16 +1350,13 @@ public void OpenRyujinxFolder()
             {
-                logPath = LoggerModule.LogDirectoryPath;
-            }
+                new DirectoryInfo(logPath).Create();
```
This doesn't need exception handling like the Logs folder creation code on Ui.Common? Maybe it should do that there?

gdkchan replied:
BTW I just noticed it uses `Directory.CreateDirectory` there, so not sure the directory is created in a different way here.

jcm93 replied:
Hmm do we even need to create a directory here? `OpenHelper` takes the path and checks if the dir exists before opening it, `AppDataManager` should have already created the directory on launch, and I don't think there's any mechanism that would change the log directory while open.

gdkchan replied:
I assume its for the case where the user deleted the folder manually for some reason.

jcm93 replied:
Alrighty, I'll just add a try/catch to OpenHelper and remove the create call here and GTK I think then.

gdkchan commented on src/Ryujinx.Ui.Common/Helper/OpenHelper.cs:
```diff
@@ -19,6 +19,14 @@ public static partial class OpenHelper
 
         public static void OpenFolder(string path)
         {
+            try
```
I'd prefer if this did not try to create the directory since its supposed to be a general helper, and in other cases where it is used, this might not be desirable.
My suggestion was more along the lines of replacing `AppDataManager.LogsDirPath` accesss on `OpenLogsFolder` with something like `AppDataManager.GetOrCreateLogsPath()` that would create the path if needed, log messages on error, and return the path (which would be null/empty if it doesn't exist and was not possible to create). Such method could also be used in other places where it employs similar logic to create the path if it doesn't exist.

jcm93 replied:
Ah, my mistake, I wasn't paying proper attention to `OpenHelper`. Will address soon, thanks.

jcm93 replied:
Fixed.

gdkchan reviewed (approved):
lgtm, thanks.

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+                catch
+                {
+                    Logger.Warning?.Print(LogClass.Application, $"Logging directory could not be created '{logDir}'");
+                    return null;
```
```suggestion

                    return null;
```

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+                        catch
+                        {
+                            Logger.Warning?.Print(LogClass.Application, $"Logging directory could not be created '{logDir}'");
+                            return null;
```
```suggestion

                            return null;
```

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+
+                    if (string.IsNullOrEmpty(logDir))
+                    {
+                        // Should evaluate to ~/Library/Application Support/Ryujinx/Logs
```
```suggestion
                        // NOTE: Should evaluate to "~/Library/Application Support/Ryujinx/Logs".
```

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+                }
+                else if (OperatingSystem.IsWindows())
+                {
+                    // Should evaluate to a 'Logs' directory in whatever directory Ryujinx was launched from.
```
```suggestion
                    // NOTE: Should evaluate to a "Logs" directory in whatever directory Ryujinx was launched from.
```

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+
+                    if (string.IsNullOrEmpty(logDir))
+                    {
+                        // Should evaluate to C:\Users\user\AppData\Roaming\Ryujinx\Logs
```
```suggestion
                        // NOTE: Should evaluate to "C:\Users\user\AppData\Roaming\Ryujinx\Logs".
```

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+                        catch
+                        {
+                            Logger.Warning?.Print(LogClass.Application, $"Logging directory could not be created '{logDir}'");
+                            return null;
```
```suggestion

                            return null;
```

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+                }
+                else if (OperatingSystem.IsLinux())
+                {
+                    // Should evaluate to ~/.config/Ryujinx/Logs
```
```suggestion
                    // NOTE: Should evaluate to "~/.config/Ryujinx/Logs".
```

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,120 @@ public static void Initialize(string baseDirPath)
+                    catch
+                    {
+                        Logger.Warning?.Print(LogClass.Application, $"Logging directory could not be created '{logDir}'");
+                        return null;
```
```suggestion

                        return null;
```

jcm93 commented:
@AcK77 addressed.

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,124 @@ public static void Initialize(string baseDirPath)
+            {
+                return LogsDirPath;
+            }
+            Logger.Notice.Print(LogClass.Application, "Logging directory not found; attempting to create new logging directory.");
+            LogsDirPath = SetUpLogsDir();
+            return LogsDirPath;
```
```suggestion

            Logger.Notice.Print(LogClass.Application, "Logging directory not found; attempting to create new logging directory.");
            LogsDirPath = SetUpLogsDir();

            return LogsDirPath;
```

jcm93 replied:
fixed.

AcK77 reviewed (commented):
I miss this one, sorry ^^'

AcK77 commented on src/Ryujinx.Common/Configuration/AppDataManager.cs:
```diff
@@ -118,9 +112,125 @@ public static void Initialize(string baseDirPath)
+            if (Directory.Exists(LogsDirPath))
+            {
+                return LogsDirPath;
+            }
+            Logger.Notice.Print(LogClass.Application, "Logging directory not found; attempting to create new logging directory.");
```
```suggestion
            }

            Logger.Notice.Print(LogClass.Application, "Logging directory not found; attempting to create new logging directory.");
```

You missed the extra empty line after the `{`. (It's just for consistencies)
